util/src/FileSystem.cpp

```diff
@@ -0,0 +1,332 @@
+bool not_found_win32_error(int errval)
+{
+  return errval == ERROR_FILE_NOT_FOUND
+      || errval == ERROR_PATH_NOT_FOUND
+      || errval == ERROR_INVALID_NAME       // "//foo"
+      || errval == ERROR_INVALID_DRIVE      // USB card reader with no card inserted
+      || errval == ERROR_NOT_READY          // CD/DVD drive with no disc inserted
+      || errval == ERROR_INVALID_PARAMETER  // ":sys:stat.h"
+      || errval == ERROR_BAD_PATHNAME       // "//nosuch" on Win64
+      || errval == ERROR_BAD_NETPATH;       // "//nosuch" on Win32
+}
+#endif
+
+#ifdef US_PLATFORM_WINDOWS
+const char DIR_SEP = DIR_SEP_WIN32;
+#else
+const char DIR_SEP = DIR_SEP_POSIX;
+#endif
+
+bool not_found_c_error(int errval)
+{
+  return errval == ENOENT || errval == ENOTDIR;
+}
+
+std::vector<std::string> SplitString(const std::string& str, const std::string& delim)
+{
+  std::vector<std::string> token;
+  std::size_t b = str.find_first_not_of(delim);
+  std::size_t e = str.find_first_of(delim, b);
+  while (e > b)
+  {
+    token.emplace_back(str.substr(b, e - b));
+    b = str.find_first_not_of(delim, e);
+    e = str.find_first_of(delim, b);
+  }
+  return token;
+}
+
+#ifndef MAXPATHLEN
+#define MAXPATHLEN 1024
+#endif
+
+std::string GetExecutablePath()
+{
+  uint32_t bufsize = MAXPATHLEN;
+  std::vector<char> buf(bufsize + 1, '\0');
+
+#ifdef US_PLATFORM_WINDOWS
+  if (GetModuleFileName(nullptr, buf.data(), bufsize) == 0 || GetLastError() == ERROR_INSUFFICIENT_BUFFER)
+  {
+    throw std::runtime_error("GetModuleFileName failed" + GetLastWin32ErrorStr());
+  }
+#elif defined(US_PLATFORM_APPLE)
+  int status = _NSGetExecutablePath(buf.data(), &bufsize);
+  if (status == -1)
+  {
+    buf.assign(bufsize + 1, '\0');
+    status = _NSGetExecutablePath(buf.data(), &bufsize);
+  }
+  if (status != 0)
+  {
+     throw std::runtime_error("_NSGetExecutablePath() failed");
+  }
+  // the returned path may not be an absolute path
+#elif defined(US_PLATFORM_LINUX)
+  ssize_t len = ::readlink("/proc/self/exe", buf.data(), bufsize);
+  if (len == -1 || len == bufsize)
+  {
+    throw std::runtime_error("Could not read /proc/self/exe into buffer");
+  }
+#else
+  // 'dlsym' does not work with symbol name 'main'
+  throw std::runtime_error("GetExecutablePath failed");
+#endif
+  return buf.data();
+}
+
+std::string InitCurrentWorkingDirectory()
+{
+#ifdef US_PLATFORM_WINDOWS
+  DWORD bufSize = ::GetCurrentDirectoryA(0, NULL);
+  if (bufSize == 0) bufSize = 1;
+  std::vector<char> buf(bufSize, '\0');
+  if (::GetCurrentDirectoryA(bufSize, buf.data()) != 0)
+  {
+    return std::string(buf.data());
+  }
+#else
+  std::size_t bufSize = PATH_MAX;
+  for(;; bufSize *= 2)
+  {
+    std::vector<char> buf(bufSize, '\0');
+    errno = 0;
+    if (getcwd(buf.data(), bufSize) != 0 && errno != ERANGE)
+    {
+      return std::string(buf.data());
+    }
+  }
+#endif
+  return std::string();
+}
+
+static const std::string s_CurrWorkingDir = InitCurrentWorkingDirectory();
```

Comment:
Why not make this magic static inside the function?

Reply:
This code was just moved from a different file and not changed. We had a discussion about the actual code in the original PR. Basically, we need to call `InitCurrentWorkingDirectory` at a time when there is only the main thread running (during static init).

Reply:
@karthikreddy09 take a look at PR #216 for more information.

Reply:
Ah ... now I understand. Thanks for pointing to the previous discussion.